contracts/utils/types/Time.sol

```diff
@@ -0,0 +1,130 @@
+     * future. The "effect" timepoint describes when the transitions happens from the "old" value to the "new" value.
+     * This allows updating the delay applied to some operation while keeping so guarantees.
+     *
+     * In particular, the {update} function guarantees that is the delay is reduced, the old delay still applies for
+     * some time. For example if the delay is currently 7 days to do an upgrade, the admin should not be able to set
+     * the delay to 0 and upgrade immediatly. If the admin wants to reduce the delay, the old delay (7 days) should
+     * still apply for some time.
+     *
+     *
+     * The `Delay` type is 128 bits long, and packs the following:
+     * [000:031] uint32 for the current value (duration)
+     * [032:063] uint32 for the pending value (duration)
+     * [064:111] uint48 for the effect date (timepoint)
+     *
+     * NOTE: The {get} and {update} function operate using timestamps. Block number based delays should use the
+     * {getAt} and {updateAt} variants of these functions.
+     */
+    type Delay is uint112;
+
+    /**
+     * @dev Wrap a Duration into a Delay to add the one-step "update in the future" feature
+     */
+    function toDelay(uint32 duration) internal pure returns (Delay) {
+        return Delay.wrap(duration);
+    }
+
+    /**
+     * @dev Get the value the Delay will be at a given timepoint
+     */
+    function getAt(Delay self, uint48 timepoint) internal pure returns (uint32) {
+        (uint32 oldValue, uint32 newValue, uint48 effect) = self.split();
+        return (effect == 0 || effect > timepoint) ? oldValue : newValue;
+    }
+
+    /**
+     * @dev Get the current value.
+     */
+    function get(Delay self) internal view returns (uint32) {
+        return self.getAt(timestamp());
+    }
+
+    /**
+     * @dev Get the pending value, and effect timepoint. If the effect timepoint is 0, then the pending value should
+     * not be considered.
+     */
+    function getPending(Delay self) internal pure returns (uint32, uint48) {
+        (, uint32 newValue, uint48 effect) = self.split();
+        return (newValue, effect);
+    }
+
+    /**
+     * @dev Update a Delay object so that a new duration takes effect at a given timepoint.
+     */
+    function updateAt(Delay self, uint32 newValue, uint48 effect) internal view returns (Delay) {
+        return pack(self.get(), newValue, effect);
+    }
+
+    /**
+     * @dev Update a Delay object so that it takes a new duration after at a timepoint that is automatically computed
+     * to enforce the old delay at the moment of the update.
+     */
+    function update(Delay self, uint32 newValue) internal view returns (Delay) {
+        uint32 value = self.get();
+        uint32 setback = value > newValue ? value - newValue : 0; // todo: 0 means immediate update. ACDAR does something more opinionated
```

Comment:
Perhaps there should be a `minEffectDelay` argument to replace 0.

Reply:
Perhaps. That would definitelly be on the "opinionated" side. Particularly in a context where this lib could be used with any clock. What should the value be ?

Reply:
My proposal was to define this function as `update(Delay self, uint32 newValue, uint32 minEffectDelay)` -- perhaps `minSetback` is a better name given the naming here. It would be used with `Math.min` against the current setback expression. It seems strictly less opinionated than the current code which hardcodes 0 as the minimum.

We wouldn't choose the value, not in this libary.

Reply:
My proposal was to define this function as `update(Delay self, uint32 newValue, uint32 minEffectDelay)` -- perhaps `minSetback` is a better name given the naming here. It would be used with `Math.min` against the current setback expression. It seems strictly less opinionated than the current code which hardcodes 0 as the minimum.

We wouldn't choose the value, not in this libary.